### ISSUE TITLE
SSL/TLS trust issues: unable to add custom certificates or disable verification

### ISSUE THREAD

**Issue by yonkeltron** [earthly/earthly#796]
Hi, I'm working in a corporate environment where I need to specify the certificates which should be used for verification of remote connections. Unfortunately, right now I can't pull from internal registries since I get the following error:

```
...snip...
     r/r/ubi8:latest | --> Load metadata linux/amd64
     r/r/ubi8:latest | WARN: (Load metadata linux/amd64) failed to do request: Head https://REGISTRY/PATH: x509: certificate signed by unknown authority
Error: failed to do request: Head https://REGISTRY/PATH: x509: certificate signed by unknown authority
```

Since the registry is contacted every time regardless of local image presence (outlined further here in #345), it appears that I have a blocker to adopting Earthly.

How can I provide more information to troubleshooting this? I would be happy to try some things out, as well. I really appreciate all of your time and efforts.

Thanks,
+Jonathan

**Comment by agbell** [comment]
Hi @yonkeltron,
One thing you could try to get around this issue, for now, is to extend the earthly/buildkit image to add in your certificate. There are [instructions for this ](https://blog.earthly.dev/mitmproxy/) here under "Extending an Alpine Image to Add a Certificate Authority". After creating that image, start it up and use the `--buildkit-host=` flag on earthly to specify its usage.  

BuildKit also has config options for this found [here](https://github.com/moby/buildkit/blob/master/docs/buildkitd.toml.md) and you could extend the image or possibly volume mount in a version of /etc/buildkitd.toml that sets things as insecure.

I think we need to provide a better way to expose this though.

**Comment by vladaionescu** [comment]
We should implement proper support for this. This could be an Earthly config item. Earthly can then automatically configure buildkit and also possibly mount any necessary cert files into the buildkit container.

**Comment by vladaionescu** [comment]
Hi @yonkeltron - the next release of Earthly will have support for this via the configuration shown in https://github.com/earthly/earthly/pull/797. Let me know if this works for you once you give it a try.

> Usage example:
> 
> ```yaml
> global:
>   buildkit_additional_args: ["-v", "<absolute-path-to-ca-file>:/etc/config/add.ca"]
>   buildkit_additional_config: |
>     [registry."<registry-hostname>"]
>       ca=["/etc/config/add.ca"]
> ```

Note that you will also be able to use an insecure registry

```yaml
global:
  buildkit_additional_config: |
    [registry."<registry-hostname>"]
      http = true
      insecure = true
```

and use `SAVE IMAGE --insecure <registry-hostname>/<image:tag>` in Earthfiles. It doesn't sound like your use-case but I wrote this here in case someone finds this on Google.

**Comment by yonkeltron** [comment]
Amazing! Can't wait to try it out and report back for you.

Thanks,
+Jonathan

**Comment by yonkeltron** [comment]
I updated this morning using Homebrew on macOS and tried both the cert specification options above and it looks like i'm not able to get the custom cert to work.

Here's my config in `~/.earthly/config.yml`

```
global:
  buildkit_additional_args:
    ["-v", "/Users/USER/Downloads/bundle.pem:/etc/config/add.ca"]
  buildkit_additional_config: |
    [registry."<registry-hostname>"]
      http = true
      insecure = true
```

Am I doing something wrong? How can I help troubleshoot? I SO appreciate your attentiveness and responsiveness.

Thanks,
+Jonathan

**Comment by vladaionescu** [comment]
Try this:

```yaml
global:
  buildkit_additional_args:
    ["-v", "/Users/USER/Downloads/bundle.pem:/etc/config/add.ca"]
  buildkit_additional_config: |
    [registry."<registry-hostname>"]
      ca=["/etc/config/add.ca"]
```

(and also replace `<registry-hostname>` with the right value)

`http` and `insecure` are only meant for unencrypted connection - they won't help in your case.

**Comment by yonkeltron** [comment]
This makes so much more sense, kindly forgive my initial misunderstanding. Unfortunately, something still might be wrong with what I'm doing.

Here's my config:

```yaml
global:
  buildkit_additional_args:
    ["-v", "/Users/USER/Downloads/bundle.pem:/etc/config/bundle.pem"]
  buildkit_additional_config: |
    [registry."registry.COMPANY.com"]
      ca=["/etc/config/bundle.pem"]

```

And here's the error I am now getting:

```
Error: failed to read /etc/config/bundle.pem: read /etc/config/bundle.pem: is a directory
```

I am going to continue playing around with setting on my end but if there is anything obviously wrong, please let me know. Also yes I am sure that `bundle.pem` is an actual PEM bundle and not a directory, so I think something about the error message is rather misleading, also.

Happy to continue trying things in service of this feature!

Thanks,
+Jonathan

**Comment by vladaionescu** [comment]
That's really strange... Is there anything relevant in the buildkit logs? `docker logs earthly-buildkitd`

`read /etc/config/bundle.pem: is a directory` is definitely a weird error.

**Comment by vladaionescu** [comment]
From your `/Users/...` path, you're probably using a mac, right? I'll test on mac on my end too, to make sure the feature works correctly.

**Comment by vladaionescu** [comment]
In parallel, can you also exec into the buildkitd container and see if you can actually read the contents of the mounted file?

```
docker exec -ti earthly-buildkitd /bin/sh
```

and then look for `/etc/config/bundle.pem` and check that it contains what you expect.

**Comment by vladaionescu** [comment]
I just tested this on my mac and I didn't encounter any issues.

(If you're feeling like it, you could run our automated test on your machine)

```
git clone git@github.com:earthly/earthly.git
cd earthly
./earthly +for-darwin
earthly=../../../build/darwin/amd64/earthly ./examples/tests/registry-certs/test.sh
```

**Comment by yonkeltron** [comment]
Snippet from logs from when I tried `/tmp/bundle.pem` instead of `/etc/config/bundle.pem`:

```
time="2021-02-25T19:54:16Z" level=error msg="/moby.buildkit.v1.frontend.LLBBridge/ResolveImageConfig returned error: rpc error: code = Unknown desc = failed to read /tmp/bundle.pem: read /tmp/bundle.pem: is a directory\n"
time="2021-02-25T19:54:16Z" level=error msg="/moby.buildkit.v1.Control/Solve returned error: rpc error: code = Unknown desc = failed to read /tmp/bundle.pem: read /tmp/bundle.pem: is a directory\n"
```

Also from when I did a `docker exec` as per the previous comment, it does look like `bundle.pem` is being created as a directory for some reason!

```bash
/ # ls -la /tmp/
total 12
drwxrwxrwt    1 root     root          4096 Feb 25 19:54 .
drwxr-xr-x    1 root     root          4096 Feb 25 19:54 ..
drwxr-xr-x    2 root     root            64 Feb 25 14:24 bundle.pem
drwxr-xr-x    4 root     root          4096 Feb 25 19:54 earthly
/ # ls -la /tmp/bundle.pem/
total 4
drwxr-xr-x    2 root     root            64 Feb 25 14:24 .
drwxrwxrwt    1 root     root          4096 Feb 25 19:54 ..
```

As for the test suite mentioned subsequently, I'll get to it ASAP!

**Comment by yonkeltron** [comment]
Looks like I *cannot* run the test suite due to SSL issues!

```
earthly-prerelease version prerelease-bd810b05130ddaa42c575c248c1db7a2b7c6e1d2 darwin/amd64
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
           buildkitd | Newer image available. Restarting buildkit daemon...
           buildkitd | ...Done
No ssh auth socket detected or zero keys loaded; falling back to https for auto auth values
golang:1.16-rc-alpine3.13 | --> Load metadata linux/amd64
            internal | --> GIT CLONE https://github.com/earthly/buildkit.git
            internal | WARN: (GIT CLONE https://github.com/earthly/buildkit.git) failed to fetch remote https://github.com/earthly/buildkit.git: exit status 128
            internal | --> docker-image://docker.io/alpine/git:v2.24.1
Error: build target: build main: bkClient.Build: failed to solve: rpc error: code = Unknown desc = failed to load cache key: failed to fetch remote https://github.com/earthly/buildkit.git: exit status 128
Check your git auth settings.
```

At least I assume it's SSL/TLS issues stemming from my environment, really.

Thanks,
+Jonathan

**Comment by vladaionescu** [comment]
Ok it's clearer now. The fact that docker mounts it as a directory seems like the core issue here.

The part `["-v", "/Users/USER/Downloads/bundle.pem:/etc/config/bundle.pem"]` simply tells earthly to pass those arguments to the `docker run` command it uses to start buildkit. Maybe it's an issue with your docker installation?

Perhaps you can try something like

```
docker run --rm -it -v /Users/USER/Downloads/bundle.pem:/tmp/bundle.pem alpine ls -la /tmp/bundle.pem
```

and see if that mounts it as a non-dir in a plain alpine image.

**Comment by vladaionescu** [comment]
I'll be around on our Slack tomorrow if you'd like to talk a bit more interactively: https://earthly.dev/slack

**Comment by yonkeltron** [comment]
Holy monkey, that was it. You're all good and I just wanted to say thanks for helping me look at this enough to find the problem!

Take it easy!

Thanks,
+Jonathan

**Comment by adamgordonbell** [comment]
I think this is better documented now:
https://docs.earthly.dev/docs/guides/configuring-registries/self-signed

**Comment by yonkeltron** [comment]
Hey @adamgordonbell and @vladaionescu !

Just putting a note here that I actually get an error from github.com when trying to run a hello world. Looks like my global git SSL/TLS settings are not being honored for some reason?

Error output:

```text
Error: build target: build main: failed to solve: failed to load cache key: error fetching default branch for repository https://github.com/earthly/hello-world.git: exit status 128
```

Should I open a new issue?

Thanks,
+Jonathan

**Comment by alexcb** [comment]
> Looks like my global git SSL/TLS settings are not being honored for some reason?

Are you still running with a config such as https://github.com/earthly/earthly/issues/796#issuecomment-786116009 ? Or can you share an example if it's changed?

Could it be that your custom pems do not contain a root level cert for github?

Perhaps you can try running

```
docker exec earthly-buildkitd /bin/sh -c 'wget -S -O /dev/null https://github.com'
```

to see if wget works within the same container?

**Comment by yonkeltron** [comment]
@alexcb hey I am in a corporate environment which requires special certificates and config.

I ran your suggested command and it did fail as hypothesized:

```bash
$ docker exec earthly-buildkitd /bin/sh -c 'wget -S -O /dev/null https://github.com'
Connecting to github.com (140.82.112.3:443)
ssl_client: github.com: certificate verification failed: unable to get local issuer certificate
wget: error getting response: Connection reset by peer
```

Does that tell us anything helpful at this point?

Thanks,
+Jonathan
